range_queries/heavy_light_decomposition.cpp

```diff
@@ -0,0 +1,576 @@
+			if(a == -1) break;
+		}
+		return ret;
+	}
+public:
+	/**
+	 * @brief Class parameterized constructor. Resizes the and initilizes the data members.
+	 * @param nodes the total number of nodes in the tree
+	*/
+	explicit HLD<X>(int nodes) : Tree<X>(nodes), SG<X>(nodes) { 
+		/* Initialization and resize vectors */
+		label = 0;
+		h_label.assign(Tree<X>::t_nodes, -1);
+		h_heavychlid.assign(Tree<X>::t_nodes, -1);
+		h_parent.resize(Tree<X>::t_nodes);
+		iota(h_parent.begin(), h_parent.end(), 0);
+	}
+
+	/**
+	 * @brief This function must be called after the tree adjacency list and node values are populated
+	 * The function initializes the required parametes, and populates the segment tree
+	*/
+	void init() {
+		Tree<X>::init();
+
+		/**
+		 * Fill the heavy child, greatest parent, and labels
+		*/
+		label = 0;
+		dfs_hc(Tree<X>::t_root);
+		dfs_par(Tree<X>::t_root);
+		dfs_labels(Tree<X>::t_root);
+
+		/**
+		 * Segment Tree Initialization
+		*/
+		for(int i = 0; i < Tree<X>::t_nodes; i++) {
+			SG<X>::s_tree[h_label[i] + Tree<X>::t_nodes] = Tree<X>::t_val[i];
+		}
+		for(int i = Tree<X>::t_nodes - 1; i > 0; i--) {
+			SG<X>::s_tree[i] = SG<X>::combine(SG<X>::s_tree[i<<1], SG<X>::s_tree[i<<1|1]);
+		}
+	}
+
+	/**
+	 * @brief This function updates the value at node with val
+	 * @param node the node where the update is done
+	 * @param val the value that is being updated
+	*/
+	void update(int node, X val) {
+		X diff  = val - Tree<X>::t_val[node];
+		SG<X>::update(h_label[node], diff);
+		Tree<X>::t_val[node] = val;
+	}
+
+	/**
+	 * @brief This function returns the sum of node values in the simple path from from node_1 to node_2
+	 * @param a the node where the simple path starts
+	 * @param b the node where the simple path ends
+	 * (parameters are interchangeable, i.e., the function is commutative)
+	 * @return the sum of node values in the simple path from a to b
+	*/
+	X query(int a, int b) {
+		int lc = Tree<X>::lca(a, b);
+		X ret = SG<X>::sret_init;
+		assert(lc != -1);
+		ret += chain_query(a, lc);
+		ret += chain_query(b, lc);
+		return ret - Tree<X>::t_val[lc];
+	}
+};
+}  // namespace range_queries
+
+/**
+ * Test implementations
+*/
+void test_1() {
+	std::cout << "Test 1:\n";
+	/**
+	 * Test details
+	*/
+	int n = 5;
+	std::vector<int64_t> node_values = {4, 2, 5, 2, 1};
+	std::vector<std::vector<int>> edges = {
+		{1, 2},
+		{1, 3},
+		{3, 4},
+		{3, 5}
+	};
+	std::vector<std::vector<int>> queries = {
+		{2, 4},
+		{1, 3, 2},
+		{2, 4},
+	};
+	
+	range_queries::HLD<int64_t> hld(n);
+	hld.set_node_val(node_values);
+	for(int i = 0; i < n - 1; i++) {
+		int u = edges[i][0], v = edges[i][1];
+		hld.add_edge(u-1, v-1);
+	}
+	hld.init();
+	for(const auto& q : queries) {
+		int type = q[0];
+		if(type == 1) {
+			int p = q[1], x = q[2];
+			hld.update(p - 1, x);
+		}
+		else if(type == 2) {
+			int p = q[1];
+			std::cout << hld.query(0, p - 1) << std::endl;
+		}
+		else {
+			continue;
+		}
+	}
+}
+
+void test_2() {
```

Comment:
This is to ensure the auto-documentation system registers this function.
```suggestion
/**
 * Second test implementations
 * @returns void
 */
static void test_2() {
```